src/OpticalFlowEstimator.cpp

```diff
@@ -0,0 +1,609 @@
+    ori_msg.data.pitch = pitch;
+    ori_msg.data.roll = roll;
+    ori_msg.data.yaw = yaw;
+    orientation_pub_.publish(ori_msg);
+
+    // Distance from the camera to the ground plane, along the camera's +z axis
+    double distance_to_plane = current_altitude_
+                             * std::sqrt(1.0
+                                       + std::pow(std::tan(pitch), 2.0)
+                                       + std::pow(std::tan(roll), 2.0));
+
+    // Converts measurements in pixels to measurements in meters in the plane
+    // parallel to the camera's xy plane and going through the point Pg, where
+    // Pg is intersection between the camera's +z axis and the ground plane
+    double current_meters_per_px = distance_to_plane
+                         / getFocalLength(target_size_,
+                                          flow_estimator_settings_.fov);
+
+    // Calculate the average velocity in the level camera frame (i.e. camera
+    // frame if our pitch and roll were zero)
+    // Note that cos(roll) is negative, because the z axis is down
+    cv::Point2f velocity_uncorrected;
+    velocity_uncorrected.x = (-average_vec.x * current_meters_per_px)
+                           / std::cos(pitch)
+                           / dt;
+    velocity_uncorrected.y = (-average_vec.y * current_meters_per_px)
+                           / -std::cos(roll)
+                           / dt;
+
+    double dp;
+    double dr;
+
+    if (last_pitch > CV_PI/2 && pitch < -CV_PI/2) {
+        dp = (pitch + 2*CV_PI - last_pitch);
+    } else if (last_pitch < -CV_PI/2 && pitch > CV_PI/2) {
+        dp = (pitch - last_pitch - 2*CV_PI);
+    } else {
+        dp = (pitch - last_pitch);
+    }
+
+    if (last_roll > CV_PI/2 && roll < -CV_PI/2) {
+        dr = (roll + 2*CV_PI - last_roll);
+    } else if (last_roll < -CV_PI/2 && roll > CV_PI/2) {
+        dr = (roll - last_roll - 2*CV_PI);
+    } else {
+        dr = (roll - last_roll);
+    }
+
+    double dpitch_dt = dp / dt;
+    double droll_dt = dr / dt;
+
+    // Observed velocity in camera frame due to rotation
+    //
+    // Positive dpdt means that the x vector rotates down, which means we think
+    // we're moving in the -x direction
+    //
+    // Positive drdt means that the y vector rotates down, which means we think
+    // we're moving in the -y direction
+    //
+    // Note again that cos(roll) is negative, because the z axis is down
+    cv::Point2f correction_vel;
+    correction_vel.x = distance_to_plane
+                     * -dpitch_dt
+                     / std::cos(pitch);
+    correction_vel.y = distance_to_plane
+                     * -droll_dt
+                     / -std::cos(roll);
+
+    // Actual velocity in level camera frame (i.e. camera frame if our pitch
+    // and roll were zero)
+    Eigen::Vector3d corrected_vel(
+        velocity_uncorrected.x - correction_vel.x,
+        velocity_uncorrected.y - correction_vel.y,
+        0.0);
+
+    last_pitch = pitch;
```

Comment:
No need to do this. These last_pitch, last_roll, pitch, and roll are method variables and are never used again.

Reply:
Fixed